packages/flutter/lib/src/rendering/mouse_tracking.dart

```diff
@@ -203,28 +341,126 @@ class MouseTracker extends ChangeNotifier {
+      });
+    }
+  }
+
+  /// Whether or not a mouse is connected and has produced events.
+  bool get mouseIsConnected => _mouseStates.isNotEmpty;
+
   // Tracks the state of connected mouse devices.
   //
   // It is the source of truth for the list of connected mouse devices.
   final Map<int, _MouseState> _mouseStates = <int, _MouseState>{};
 
+  // Used to wrap any procedure that might change `mouseIsConnected`.
+  //
+  // This method records `mouseIsConnected`, runs `task`, and calls
+  // [notifyListeners] at the end if the `mouseIsConnected` has changed.
+  void _monitorMouseConnection(VoidCallback task) {
+    final bool mouseWasConnected = mouseIsConnected;
+    task();
+    if (mouseWasConnected != mouseIsConnected)
+      notifyListeners();
+  }
+
+  bool _debugDuringDeviceUpdate = false;
+  // Used to wrap any procedure that might call [handleDeviceUpdate].
+  //
+  // In debug mode, this method uses `_debugDuringDeviceUpdate` to prevent
+  // `_deviceUpdatePhase` being recursively called.
+  void _deviceUpdatePhase(VoidCallback task) {
+    assert(!_debugDuringDeviceUpdate);
+    assert(() {
+      _debugDuringDeviceUpdate = true;
+      return true;
+    }());
+    task();
+    assert(() {
+      _debugDuringDeviceUpdate = false;
+      return true;
+    }());
+  }
+
   // Whether an observed event might update a device.
-  static bool _shouldMarkStateDirty(_MouseState state, PointerEvent value) {
+  static bool _shouldMarkStateDirty(_MouseState state, PointerEvent event) {
     if (state == null)
       return true;
-    assert(value != null);
+    assert(event != null);
     final PointerEvent lastEvent = state.latestEvent;
-    assert(value.device == lastEvent.device);
+    assert(event.device == lastEvent.device);
     // An Added can only follow a Removed, and a Removed can only be followed
     // by an Added.
-    assert((value is PointerAddedEvent) == (lastEvent is PointerRemovedEvent));
+    assert((event is PointerAddedEvent) == (lastEvent is PointerRemovedEvent));
 
     // Ignore events that are unrelated to mouse tracking.
-    if (value is PointerSignalEvent)
+    if (event is PointerSignalEvent)
       return false;
     return lastEvent is PointerAddedEvent
-      || value is PointerRemovedEvent
-      || lastEvent.position != value.position;
+      || event is PointerRemovedEvent
+      || lastEvent.position != event.position;
+  }
+
+  // Find the annotations that is hovered by the device of the `state`.
+  //
+  // If the device is not connected, an empty set is returned without calling
+  // `annotationFinder`.
+  LinkedHashSet<MouseTrackerAnnotation> _findAnnotations(_MouseState state) {
+    final Offset globalPosition = state.latestEvent.position;
+    final int device = state.device;
+    return (_mouseStates.containsKey(device))
+      ? LinkedHashSet<MouseTrackerAnnotation>.from(annotationFinder(globalPosition))
+      : <MouseTrackerAnnotation>{} as LinkedHashSet<MouseTrackerAnnotation>;
```

Comment:
avoid the cast by just instantiating it directly?

```suggestion
      : LinkedHashSet<MouseTrackerAnnotation>();
```

Reply:
You could potentially also mark it const? (Assuming nothing gets ever added to it)

Reply:
I would like this change, but the linter keeps alerting "Use collective literals when possible".